### ISSUE TITLE
Checkboxes and async loading of children

### ISSUE THREAD

**Issue by wendelstam** [500tech/angular-tree-component#531]
Hi,

I am loading a tree with multiple levels with checkboxes (+4000 nodes) and I use async loading.
First I got an error but after monkey patching the method setIsSelected it works.

```
TreeNode.prototype.setIsSelected = function (value) {
  if (this.isLeaf || !this.children) {
    this.treeModel.setSelectedNode(this, value);
  } else {
    this.treeModel.setSelectedNode(this, value);
    this.children.forEach((child) => child.setIsSelected(value));
  }
  return this;
};
```

But when I'm trying to check only node paths fully loaded gets selected. Any workaround for this?

**UPDATE**

Was able to make it work by adding this.

```
Object.defineProperty(TreeNode.prototype, 'isSelected', {
  get: function () {
    if (this.isLeaf || !this.children) {
      return this.treeModel.isSelected(this);
    } else {
      return some(this.children, function (node) {
        return node.isSelected;
      });
    }
  },
  enumerable: true,
  configurable: true
});
```

PS:
Great component all and all!

**Comment by wendelstam** [comment]
When trying to check checkboxes through the API its quite slow, have tried to set selectedLeafNodeIds directly on the state object but cannot get the check state updated/propagated into the tree. Any ideas? Same for expandedNodeIds. What am I missing?

**Comment by klteograss** [comment]
Yes, using `node.setIsSelected()` is too slow, especially when you have multiple checkbox selected.
Is there any faster way of initialize the checkbox states?

**Comment by isaklafleur** [comment]
@wendelstam @klteograss  maybe this will solve the issue?

I just made a [pull ](https://github.com/500tech/angular-tree-component/pull/547)request as the api does not export `selectedLeafNodeIds` in the `ITreeState` interface.

**Comment by klteograss** [comment]
@isaklafleur thanks, yes.  I solved the performance issue by using the selectedLeafNodeIds in the tree model.

**Comment by omid-nazifi** [comment]
Hi @wendelstam, Could you tell me how you have made monkey patching for setIsSelected method? Where should I replace the method?

**Comment by wendelstam** [comment]
You should just be able to include it anywhere so it is just executed prior to you using the tree component @omid-nazifi, I think I just included it before me component class when I tested it.
(Have since then moved to using Teleriks Treeview component i.e. a paid alternative). Still really like this project though.

**Comment by omid-nazifi** [comment]
Sorry, I don't understand where is it exactly. Did you add it out of your component? How does it call? I mean how does app recognize it instead of original method?
Could you display me by example,  If it is possible?

**Comment by wendelstam** [comment]
@omid-nazifi dont have the code left but I just added it before to the component class definition i.e. like a "global" const (outside the class) so it is just defined once.
In production i would probably have refactored it so it was in a separate ts/js file loaded individually by ng-cli/webpack. But as long as it's defined after the tree component is loaded I think you should be good. Did not test it alot since I found other things for our project that required a different tree view component (lack of some for us important features not bugs).

```
Object.defineProperty(TreeNode.prototype, 'isSelected', {
  get: function () {
    if (this.isLeaf || !this.children) {
      return this.treeModel.isSelected(this);
    } else {
      return some(this.children, function (node) {
        return node.isSelected;
      });
    }
  },
  enumerable: true,
  configurable: true
});

@Component({...})
export class MyComponent {
....
}
```

**Comment by omid-nazifi** [comment]
Thanks you @wendelstam 
It was a great note. :-)

**Comment by omid-nazifi** [comment]
Hi @isaklafleur @klteograss,
Could you help me how you solved low performance by selectedLeafNodeIds when checkbox is selected?

Thank you

**Comment by phhbr** [comment]
Did you try to work with the tri-state checkbox aswell?

Cannot get it to work properly, and don't have any clue why it does not work

**Comment by adamkleingit** [comment]
I hope after I merge your PR it will help - I made a commit for you to
review. Thanks

**Comment by adamkleingit** [comment]
Published in 8.0.2
